manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,639 @@
+        //If the predicate fails, then no changes need to happen.
+
+        // newValue is not 1, previousValue == 1
+        if(!(!newValueTest && previousValueTest)) {
+            return Optional.empty();
+        }
+
+        // Grab all datapoints (To be replaced by AssetDatapointValueQuery)
+        List<AssetDatapoint> valueDatapoints = AssetDatapointService.getDatapoints(ref);
+        ArrayList<AssetDatapoint> list = new ArrayList<>(valueDatapoints);
+
+        // If there are no historical data found, add some first
+        if(list.isEmpty()) return Optional.empty();
+
+        //What we do now is, we will try to figure out the latest datapoint where the predicate fails,
+        //before the newValue.
+        //This means that, the state change took place between the datapoint we just found and its next one.
+
+        //Find the first datapoint that passes the negated predicate
+
+        AssetDatapoint StateChangeAssetDatapoint = null;
+
+        try {
+            for (int i = 0; i < list.size()-1; i++) {
+                // Not using Object.equals, but Datapoint.equals
+
+                AssetDatapoint currentDp = list.get(i);
+                AssetDatapoint theVeryPreviousDp = list.get(i+1);
+
+    //            So, if the currentDp passes the predicate,
+                boolean currentDpTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(currentDp.getValue());
+    //            and if the very previous one (NEXT one in the array and PREVIOUS in the time dimension).
+    //            FAILS the predicate,
+                boolean previousDpTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(theVeryPreviousDp.getValue());
+    //            A state change happened where the state we are looking for was turned on.
+    //            We want the currentDp.
+
+
+                if(currentDpTest && !previousDpTest){
+                    StateChangeAssetDatapoint = currentDp;
+                    break;
+                }
+            }
+
+            if (StateChangeAssetDatapoint != null){
+                if (!pred.value.asPredicate(timerService::getCurrentTimeMillis).test(StateChangeAssetDatapoint.getValue())){
+                    throw new Exception("Found state change datapoint failed predicate");
+                }
+            }else{
+                throw new Exception("Couldn't find asset state change value");
+            }
+
+        }catch (Exception e){
+            getLogger().warning(e.getMessage());
+            return Optional.empty();
+        }
+
+        //BUG: The Timestamp of an asset datapoint is NOT attribute.getTimestamp()!
+        //This makes it much harder to relate datapoints to relate a duration to the time-distance between two datapoints.
+        //To make this easier, we are going to use some time-padding of 1 second from the gap. It's going to be sufficient to cover some edge-cases.
```

Comment:
Hmmmm interesting finding! Don't know if we can change this in the datapoint code.
You can create a GitHub issue for this. I guess you can also assign it to me.

Don't know how large the gap is between an datapoint timestamp and an attributeChanged timestamp?

I would hesitate to practice a "margin of error" of 1 second.
Imagine the system being slow, taking for example 500+ ms to handle an attribute change, it could hurt this.
Maybe something like querying the datapoint with the closest timestamp and the same value would work?

Reply:
Of course the margin of error was added at least so that I can get my data in, because without it I lost a lot of datapoints. It's not optimal in any way. 

Lets continue this in #1156.